File: aoc/day02_cube_contundrum/part2.py
import functools
import io
from collections import defaultdict
from itertools import chain

from .part1 import Record, parse_game

BAG = defaultdict(
    int,
    {
        "red": 12,
        "green": 13,
        "blue": 14,
    },
)


def process_color(color: str, flat_game: list[Record]) -> int:
    color_occurences = list(filter(lambda record: record.color == color, flat_game))
    if len(color_occurences) > 1:
        largest = max(*color_occurences, key=lambda record: record.count)
    else:
        largest = color_occurences[0]
    return largest.count


def process_game(line: str) -> int:
    _, game = parse_game(line)
    flat_game = list(chain.from_iterable(game))
    colors = set(map(lambda record: record.color, flat_game))
    return functools.reduce(
        lambda x, y: x * process_color(y, flat_game),
        colors,
        1,
    )


def main(data: io.TextIOWrapper) -> int:
    return sum(map(process_game, data))

File: lib/day02/part2.ex
defmodule Aoc.Day02part2 do
  @spec main(binary) :: integer
  def main(data) do
    parse_record = fn s ->
      [count, color] = String.split(s)
      {String.to_integer(count), color}
    end

    parse_set = fn s ->
      s
      |> String.split(", ")
      |> Enum.map(parse_record)
    end

    parse_game = fn s ->
      s
      |> String.split(": ")
      |> Enum.at(1)
      |> String.split("; ")
      |> Enum.map(parse_set)
      |> Enum.concat()
    end

    process_color = fn color, flat_game ->
      flat_game
      |> Enum.filter(fn record -> elem(record, 1) == color end)
      |> Enum.map(fn record -> elem(record, 0) end)
      |> Enum.max()
    end

    process_game = fn flat_game ->
      flat_game
      |> Enum.map(fn record -> elem(record, 1) end)
      |> Enum.uniq()
      |> Enum.reduce(1, fn color, acc -> acc * process_color.(color, flat_game) end)
    end

    data
    |> String.split("\n")
    |> Enum.filter(fn s -> s != "" end)
    |> Enum.map(parse_game)
    |> Enum.map(process_game)
    |> Enum.sum()
  end
end
